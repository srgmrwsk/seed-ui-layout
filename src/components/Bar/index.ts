export * from './Bar';
export { default as Bar } from './Bar';
