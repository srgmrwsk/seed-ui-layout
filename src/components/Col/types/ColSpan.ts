export type ColSpan = boolean | number | 'auto';
