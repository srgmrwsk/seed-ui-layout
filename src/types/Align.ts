export type Align = 'start' | 'end' | 'center' | 'baseline' | 'stretch';
