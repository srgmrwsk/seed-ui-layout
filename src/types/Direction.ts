export type Direction = 'column' | 'column-reverse' | 'row' | 'rew-reverse';
