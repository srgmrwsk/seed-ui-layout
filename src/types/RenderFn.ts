import React from 'react';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export type RenderFn = (component: React.ReactElement, ...args: []) => any;
