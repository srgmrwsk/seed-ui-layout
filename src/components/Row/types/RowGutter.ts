import { Sizing } from '../../../types/Sizing';

export type RowGutter = Sizing | [Sizing, Sizing];
