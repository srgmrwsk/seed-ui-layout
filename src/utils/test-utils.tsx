import React from 'react';
import { render } from '@testing-library/react';
import { renderToStaticMarkup } from 'react-dom/server';
import { configureAxe } from 'jest-axe';

import { RenderFn } from '../types/RenderFn';

export const axe = configureAxe({
  rules: {
    // disable landmark rules when testing isolated components
    region: { enabled: false },
  },
});

export function create(
  ...args: Parameters<RenderFn>
): ChildNode | HTMLCollection | null {
  const { container } = render(...args);
  return container.children.length > 1
    ? container.children
    : container.firstChild;
}
export function renderToHtml(component: React.ReactElement): string {
  return renderToStaticMarkup(component);
}
