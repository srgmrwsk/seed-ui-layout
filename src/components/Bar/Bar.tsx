import React from 'react';
import cx from 'classnames';

import { ResponsiveValue } from '../../types/ResponsiveValue';
import { getSizingVariables } from '../../utils/get-sizing-variables';
import { getOffsetRightVariables } from '../../utils/get-offset-right-variables';
import { Position } from '../../types/Position';
import { getStringStyles } from '../../utils/get-string-styles';
import { getBooleanStyles } from '../../utils/get-boolean-styles';
import { Offset } from '../../types/Offset';
import { getOffsetBottomVariables } from '../../utils/get-offset-bottom-variables';
import { getOffsetLeftVariables } from '../../utils/get-offset-left-variables';
import { getOffsetTopVariables } from '../../utils/get-offset-top-variables';
import { Sizing } from '../../types/Sizing';

export interface BarProps
  extends Omit<
    React.AllHTMLAttributes<HTMLElement>,
    'as' | 'height' | 'width'
  > {
  as?: React.ElementType;
  height?: ResponsiveValue<Sizing>;
  maxHeight?: ResponsiveValue<Sizing>;
  maxWidth?: ResponsiveValue<Sizing>;
  minHeight?: ResponsiveValue<Sizing>;
  minWidth?: ResponsiveValue<Sizing>;
  offset?: ResponsiveValue<Offset>;
  position?: ResponsiveValue<Position>;
  width?: ResponsiveValue<Sizing>;
}

function Bar(
  {
    as: As = 'div',
    className,
    height,
    maxHeight,
    maxWidth,
    minHeight,
    minWidth,
    offset,
    position,
    style,
    width,
    children,
    ...elemProps
  }: BarProps,
  ref: React.Ref<HTMLElement>,
): JSX.Element {
  return (
    <As
      {...elemProps}
      ref={ref}
      className={cx(
        getBooleanStyles('_seed-h', height),
        getBooleanStyles('_seed-max-h', maxHeight),
        getBooleanStyles('_seed-max-w', maxWidth),
        getBooleanStyles('_seed-min-h', minHeight),
        getBooleanStyles('_seed-min-w', minWidth),
        getBooleanStyles('_seed-offset', offset),
        getStringStyles('_seed-pos', position),
        getBooleanStyles('_seed-w', width),
        className,
      )}
      style={{
        ...getSizingVariables('seed-h', height),
        ...getSizingVariables('seed-max-h', maxHeight),
        ...getSizingVariables('seed-max-w', maxWidth),
        ...getSizingVariables('seed-min-h', minHeight),
        ...getSizingVariables('seed-min-w', minWidth),
        ...getOffsetBottomVariables('seed-offset-bottom', offset),
        ...getOffsetLeftVariables('seed-offset-left', offset),
        ...getOffsetRightVariables('seed-offset-right', offset),
        ...getOffsetTopVariables('seed-offset-top', offset),
        ...getSizingVariables('seed-w', width),
        ...style,
      }}
    >
      {children}
    </As>
  );
}

Bar.displayName = 'Bar';

export default React.forwardRef(Bar);
