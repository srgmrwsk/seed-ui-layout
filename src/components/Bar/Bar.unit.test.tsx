import React from 'react';
import { renderToStaticMarkup } from 'react-dom/server';

import { RenderFn } from '../../types/RenderFn';
import { axe, create } from '../../utils/test-utils';

import Bar, { BarProps } from './Bar';

describe('Bar', () => {
  const setup = (renderFn: RenderFn, props?: BarProps) =>
    renderFn(<Bar {...props} />);

  it('should render correctly with default props', () => {
    const actual = setup(create);
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with as prop', () => {
    const actual = setup(create, { as: 'header' });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with children prop', () => {
    const actual = setup(create, { children: <p>Element with text.</p> });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with bottom prop', () => {
    const actual = setup(create, { bottom: 500 });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with bottom responsive prop', () => {
    const actual = setup(create, {
      bottom: {
        xs: 'auto',
        sm: 0.75,
        md: 500,
        lg: 625,
        xl: 750,
        xxl: 875,
      },
    });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with height prop', () => {
    const actual = setup(create, { height: 500 });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with height responsive prop', () => {
    const actual = setup(create, {
      height: {
        xs: 'auto',
        sm: 0.75,
        md: 500,
        lg: 625,
        xl: 750,
        xxl: 875,
      },
    });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with left prop', () => {
    const actual = setup(create, { left: 500 });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with left responsive prop', () => {
    const actual = setup(create, {
      left: {
        xs: 'auto',
        sm: 0.75,
        md: 500,
        lg: 625,
        xl: 750,
        xxl: 875,
      },
    });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with max-height prop', () => {
    const actual = setup(create, { maxHeight: 500 });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with max-height responsive prop', () => {
    const actual = setup(create, {
      maxHeight: {
        xs: 'auto',
        sm: 0.75,
        md: 500,
        lg: 625,
        xl: 750,
        xxl: 875,
      },
    });

    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with max-width prop', () => {
    const actual = setup(create, { maxWidth: 500 });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with max-width responsive prop', () => {
    const actual = setup(create, {
      maxWidth: {
        xs: 'auto',
        sm: 0.75,
        md: 500,
        lg: 625,
        xl: 750,
        xxl: 875,
      },
    });

    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with min-height prop', () => {
    const actual = setup(create, { minHeight: 500 });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with min-height responsive prop', () => {
    const actual = setup(create, {
      minHeight: {
        xs: 'auto',
        sm: 0.75,
        md: 500,
        lg: 625,
        xl: 750,
        xxl: 875,
      },
    });

    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with min-width prop', () => {
    const actual = setup(create, { minWidth: 500 });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with min-width responsive prop', () => {
    const actual = setup(create, {
      minWidth: {
        xs: 'auto',
        sm: 0.75,
        md: 500,
        lg: 625,
        xl: 750,
        xxl: 875,
      },
    });

    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with position prop', () => {
    const actual = setup(create, { position: 'absolute' });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with position responsive prop', () => {
    const actual = setup(create, {
      position: {
        xs: 'absolute',
        sm: 'fixed',
        md: 'relative',
        lg: 'static',
        xl: 'sticky',
        xxl: 'absolute',
      },
    });

    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with right prop', () => {
    const actual = setup(create, { right: 500 });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with right responsive prop', () => {
    const actual = setup(create, {
      right: {
        xs: 'auto',
        sm: 0.75,
        md: 500,
        lg: 625,
        xl: 750,
        xxl: 875,
      },
    });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with top prop', () => {
    const actual = setup(create, { top: 500 });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with top responsive prop', () => {
    const actual = setup(create, {
      top: {
        xs: 'auto',
        sm: 0.75,
        md: 500,
        lg: 625,
        xl: 750,
        xxl: 875,
      },
    });
    expect(actual).toMatchSnapshot();
  });
  it('should render correctly with width prop', () => {
    const actual = setup(create, { width: 500 });
    expect(actual).toMatchSnapshot();
  });

  it('should render correctly with width responsive prop', () => {
    const actual = setup(create, {
      width: {
        xs: 'auto',
        sm: 0.75,
        md: 500,
        lg: 625,
        xl: 750,
        xxl: 875,
      },
    });

    expect(actual).toMatchSnapshot();
  });

  it('should meet accessibility guidelines', async () => {
    const wrapper = setup(renderToStaticMarkup);
    const actual = await axe(wrapper);
    expect(actual).toHaveNoViolations();
  });
});
