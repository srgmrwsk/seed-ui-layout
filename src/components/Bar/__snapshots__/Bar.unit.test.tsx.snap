// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bar should render correctly with as prop 1`] = `
<header
  class=""
/>
`;

exports[`Bar should render correctly with bottom prop 1`] = `
<div
  class="_seed-bottom"
  style="--seed-bottom: 500px;"
/>
`;

exports[`Bar should render correctly with bottom responsive prop 1`] = `
<div
  class="_seed-bottom _seed-bottom-sm _seed-bottom-md _seed-bottom-lg _seed-bottom-xl _seed-bottom-xxl"
  style="--seed-bottom: auto; --seed-bottom-sm: 75%; --seed-bottom-md: 500px; --seed-bottom-lg: 625px; --seed-bottom-xl: 750px; --seed-bottom-xxl: 875px;"
/>
`;

exports[`Bar should render correctly with children prop 1`] = `
<div
  class=""
>
  <p>
    Element with text.
  </p>
</div>
`;

exports[`Bar should render correctly with default props 1`] = `
<div
  class=""
/>
`;

exports[`Bar should render correctly with height prop 1`] = `
<div
  class="_seed-h"
  style="--seed-h: 500px;"
/>
`;

exports[`Bar should render correctly with height responsive prop 1`] = `
<div
  class="_seed-h _seed-h-sm _seed-h-md _seed-h-lg _seed-h-xl _seed-h-xxl"
  style="--seed-h: auto; --seed-h-sm: 75%; --seed-h-md: 500px; --seed-h-lg: 625px; --seed-h-xl: 750px; --seed-h-xxl: 875px;"
/>
`;

exports[`Bar should render correctly with left prop 1`] = `
<div
  class="_seed-left"
  style="--seed-left: 500px;"
/>
`;

exports[`Bar should render correctly with left responsive prop 1`] = `
<div
  class="_seed-left _seed-left-sm _seed-left-md _seed-left-lg _seed-left-xl _seed-left-xxl"
  style="--seed-left: auto; --seed-left-sm: 75%; --seed-left-md: 500px; --seed-left-lg: 625px; --seed-left-xl: 750px; --seed-left-xxl: 875px;"
/>
`;

exports[`Bar should render correctly with max-height prop 1`] = `
<div
  class="_seed-max-h"
  style="--seed-max-h: 500px;"
/>
`;

exports[`Bar should render correctly with max-height responsive prop 1`] = `
<div
  class="_seed-max-h _seed-max-h-sm _seed-max-h-md _seed-max-h-lg _seed-max-h-xl _seed-max-h-xxl"
  style="--seed-max-h: auto; --seed-max-h-sm: 75%; --seed-max-h-md: 500px; --seed-max-h-lg: 625px; --seed-max-h-xl: 750px; --seed-max-h-xxl: 875px;"
/>
`;

exports[`Bar should render correctly with max-width prop 1`] = `
<div
  class="_seed-max-w"
  style="--seed-max-w: 500px;"
/>
`;

exports[`Bar should render correctly with max-width responsive prop 1`] = `
<div
  class="_seed-max-w _seed-max-w-sm _seed-max-w-md _seed-max-w-lg _seed-max-w-xl _seed-max-w-xxl"
  style="--seed-max-w: auto; --seed-max-w-sm: 75%; --seed-max-w-md: 500px; --seed-max-w-lg: 625px; --seed-max-w-xl: 750px; --seed-max-w-xxl: 875px;"
/>
`;

exports[`Bar should render correctly with min-height prop 1`] = `
<div
  class="_seed-min-h"
  style="--seed-min-h: 500px;"
/>
`;

exports[`Bar should render correctly with min-height responsive prop 1`] = `
<div
  class="_seed-min-h _seed-min-h-sm _seed-min-h-md _seed-min-h-lg _seed-min-h-xl _seed-min-h-xxl"
  style="--seed-min-h: auto; --seed-min-h-sm: 75%; --seed-min-h-md: 500px; --seed-min-h-lg: 625px; --seed-min-h-xl: 750px; --seed-min-h-xxl: 875px;"
/>
`;

exports[`Bar should render correctly with min-width prop 1`] = `
<div
  class="_seed-min-w"
  style="--seed-min-w: 500px;"
/>
`;

exports[`Bar should render correctly with min-width responsive prop 1`] = `
<div
  class="_seed-min-w _seed-min-w-sm _seed-min-w-md _seed-min-w-lg _seed-min-w-xl _seed-min-w-xxl"
  style="--seed-min-w: auto; --seed-min-w-sm: 75%; --seed-min-w-md: 500px; --seed-min-w-lg: 625px; --seed-min-w-xl: 750px; --seed-min-w-xxl: 875px;"
/>
`;

exports[`Bar should render correctly with position prop 1`] = `
<div
  class="_seed-pos_absolute"
/>
`;

exports[`Bar should render correctly with position responsive prop 1`] = `
<div
  class="_seed-pos_absolute _seed-pos-sm_fixed _seed-pos-md_relative _seed-pos-lg_static _seed-pos-xl_sticky _seed-pos-xxl_absolute"
/>
`;

exports[`Bar should render correctly with right prop 1`] = `
<div
  class="_seed-right"
  style="--seed-right: 500px;"
/>
`;

exports[`Bar should render correctly with right responsive prop 1`] = `
<div
  class="_seed-right _seed-right-sm _seed-right-md _seed-right-lg _seed-right-xl _seed-right-xxl"
  style="--seed-right: auto; --seed-right-sm: 75%; --seed-right-md: 500px; --seed-right-lg: 625px; --seed-right-xl: 750px; --seed-right-xxl: 875px;"
/>
`;

exports[`Bar should render correctly with top prop 1`] = `
<div
  class="_seed-top"
  style="--seed-top: 500px;"
/>
`;

exports[`Bar should render correctly with top responsive prop 1`] = `
<div
  class="_seed-top _seed-top-sm _seed-top-md _seed-top-lg _seed-top-xl _seed-top-xxl"
  style="--seed-top: auto; --seed-top-sm: 75%; --seed-top-md: 500px; --seed-top-lg: 625px; --seed-top-xl: 750px; --seed-top-xxl: 875px;"
/>
`;

exports[`Bar should render correctly with width prop 1`] = `
<div
  class="_seed-w"
  style="--seed-w: 500px;"
/>
`;

exports[`Bar should render correctly with width responsive prop 1`] = `
<div
  class="_seed-w _seed-w-sm _seed-w-md _seed-w-lg _seed-w-xl _seed-w-xxl"
  style="--seed-w: auto; --seed-w-sm: 75%; --seed-w-md: 500px; --seed-w-lg: 625px; --seed-w-xl: 750px; --seed-w-xxl: 875px;"
/>
`;
